### PR TITLE
Refactor `Server` Library + Improve Docs

### PR DESCRIPTION
**Pull request by wolveix**
Heya!

This was the other PR I worked on last week. It just refactors some of the code within the `Server` library (focusing on the `Manager` directory), standardizes a few coding conventions, and adds docs to everything (basic docs for most functions, more thorough docs for others). I also reordered functions to sit alongside their callers, the best example being getters and setters being together.

I'd appreciate feedback! :)

### CODE REVIEW

**Comment by wolveix** [comment]
Hah, merging conflicts through the GitHub UI strikes again

**Comment by BelleNottelling** [comment]
So.. to be honest, there's not much point in a lot of these PHPDocs and many will be removed with our automated tools. (Much like [this](https://github.com/FOSSBilling/FOSSBilling/pull/2061/files#diff-5014a44a071fad477ebf24dd1004aaf00dba2fdc5d74a72f6dc28d1e7bd830a7L152-L155))

For example:
```PHP
    /**
     * @param $param
     * @return $this
     */
    public function setUsername($param): static
    {
        $this->username = $param;

        return $this;
    }
```
`@param $param` doesn't describe anything that the function itself doesn't already describe and will be removed automatically on the next Rector & PHP-CS-Fixer run.
(Also, is `static` correct? Shouldn't it be `object`?)

Here's another case. The type hint already specifies what the return value is, so the PHPDoc isn't adding anything of value and will automatically be removed.
```PHP
    /**
     * @return mixed
     */
    public function getMaxQuota(): mixed
    {
        return $this->quota;
    }
```

Realistically, the PHPDocs should only be written where they are adding in information that wasn't already there.
Here's an example of a PHPDoc that is providing additional context. The params specified in it are describing what the parameters are supposed to be and what influence they have on the behavior of the function:
```PHP
/**
 * Accepts a string and either repeats it a specified number of times or returns it directly.
 * 
 * @param string $string The string to repeat / return
 * @param null|int $numberOrNull The number of times to repeat the number. (0 or `null` will skip the repeating step)
 */
function doSomethingRandomAndPointless(string $string, ?int $numberOrNull): string
{
    // If `numberOrNull` is not falsey (so not 0 or `null` for this example), repeat the provided string and return that
    if($numberOrNull){
        return str_repeat($string, $numberOrNull);
    } else {
        return $string;
    }
}
```

Whereas in this case, the PHPDoc is only repeating what the function's type hinting tells us:
```PHP
/**
 * @param string $string 
 * @param null|int $numberOrNull 
 * @return string 
 */
function doSomethingRandomAndPointless(string $string, ?int $numberOrNull): string
{
    // If `numberOrNull` is not falsey (so not 0 or `null` for this example), repeat the provided string and return that
    if($numberOrNull){
        return str_repeat($string, $numberOrNull);
    } else {
        return $string;
    }
}
```

Additionally, I am seeing many of these which are completely wrong. For example:
```PHP
/**
 * @return null
 */
public function getUsername()
{
    return $this->username;
}
```

While the `username` property of this class starts with a value of `null`, it'll almost certainly always return a string when accessed. This can also be represented as a type hint which will better the static analysis and strengthen how typed the codebase. (Though truthfully, some old BoxBilling stuff can be a bit problematic when we mess with type-hints, but so long as it's not under the `/Library/Model` directory things are unlikely to cause an issue)

**Comment by wolveix** [comment]
Hah, blame my IDE's automated PHPDocs for the start of the meaningless ones. I'll add information to the useless ones, and remove the ones that don't really need information (like getters and setters)

**Comment by wolveix** [comment]
@BelleNottelling I ended up not removing the PHPDoc blocks for the getters and setters, as there's _now_ some additional context in many of the blocks. I'd still be happy to remove them if you decide that they're not worthwhile, even in their current state :) Many of the _wrong_ blocks were due to the model fields not having explicit type declarations, which I've now rectified. I've left most of them as `?string`s, to prevent any breaking changes anywhere.

> Also, is `static` correct? Shouldn't it be `object`?

I believe static is actually correct in this context :)

**Comment by BelleNottelling** [comment]
@wolveix Thanks, the PHPDocs are looking much better now.

The tests are failing on this PR though and that does need to addressed before it can be merged.
It looks like you just need to update the PHPStan baseline (`php src/vendor/bin/phpstan --generate-baseline`) as these all look to be already existing issues.

I'm only seeing the addition of some PHPDocs and some restructured code, are there any behavior changes that I'm not noticing?

> I previously understood this to be common practice in PHP, though from brushing up on my now-outdated knowledge of the language, it seems to be the opposite

I believe it used to be common practice, but these days you'll find most PHP projects (including ourselves) generally sticking to the [PSR-12](https://www.php-fig.org/psr/psr-12/) coding standard which actually explicitly says against it:

> Property names MUST NOT be prefixed with a single underscore to indicate protected or private visibility. That is, an underscore prefix explicitly has no meaning.

In our case, any code that does it was either written in the wrong style or is just old code 🙃

**Comment by wolveix** [comment]
Awesome! Nope, no real logic changes. I cleaned up/improved a few if statements, primarily inverted them to avoid unnecessary nesting. It looks like one of the errors is due to specifically forcing `maxBandwidth` as a string, which may cause problems with the maths in the Plesk code (L523). It's only specified as a string as DirectAdmin may fill it with `unlimited` as a string (oh what a fun server to work with...). 

> I believe it used to be common practice, but these days you'll find most PHP projects (including ourselves) generally sticking to the [PSR-12](https://www.php-fig.org/psr/psr-12/) coding standard which actually explicitly says against it:

Yeah, that's what I came across. I can configure my IDE to force hints for that standard, so great!

**Comment by wolveix** [comment]
I forgot about the `src/modules/Servicehosting` directory. I'm gonna start working on cleaning that up a bit and improving docs. It can probably wait for _yet another PR_, though.

File: src/library/Server/Manager/Whm.php
Comment (BelleNottelling) [inline src/library/Server/Manager/Whm.php:443]: Okay.. Why?
Both variables are descriptively named

File: src/library/Server/Manager/Plesk.php
Comment (BelleNottelling) [inline src/library/Server/Manager/Plesk.php:645]: Looks like these aren't supposed to be part of this PR, please remove them

**Review by BelleNottelling — CHANGES_REQUESTED** [review]
Couple of minor things I noticed.

I also remain very hesitant about about the addition of the typehints on the properties of this older code. I've been burned far too many times by unexpected behavior from BoxBilling code that breaks when we've added typehints. Have you had the chance to actually verify that it behaves correctly with those modifications?

File: src/library/Server/Manager/Plesk.php
Comment (BelleNottelling) [reply src/library/Server/Manager/Plesk.php:645]: Oh, It looks like I accidentally un-highlighted the lines in question.
It's supposed to be surrounding the DNS related functions you added which appear to be part of something else you are working on which is separate from this PR

Comment (wolveix) [reply src/library/Server/Manager/Plesk.php:645]: Hmm, which DNS functions are you referring to? There's a chance things ended up getting merged that weren't supposed to, but I'm not sure I see anything my end

File: src/library/Server/Manager/Whm.php
Comment (wolveix) [reply src/library/Server/Manager/Whm.php:443]: Whoops, that may have been before variables were renamed. Removing

Comment (wolveix) [reply src/library/Server/Manager/Whm.php:443]: Looks like there are a few occurences of this, apologies for the oversight! Removing them

**Comment by BelleNottelling** [comment]
> It can probably wait for _yet another PR_, though.

I actually have planned some (smaller) changes in these files which will likely be done today, so it would be easiest to just get this PR done and merged as my changes will be easier to rebase off yours VS the other way around.

> looks like one of the errors is due to specifically forcing `maxBandwidth` as a string, which may cause problems with the maths in the Plesk code (L523)

Ah thanks. Given the additional context, we should probably perform an `is_numeric` check and then use `intval` or `floatval` (Not sure if there'd be reason for one of these to be a float?) before performing the multiplication.

Also I'm seeing stuff that will for sure be changed with the CS & coding standards workflow. If you wanted to go above-and-beyond you can run Rector and then PHP-CS-Fixer to get those changes in now rather than waiting till the next run of the workflow to create a separate PR for it.

(It's just doing `php src/vendor/bin/rector` and then `php src/vendor/bin/php-cs-fixer fix`)

File: src/library/Server/Manager/Plesk.php
Comment (BelleNottelling) [reply src/library/Server/Manager/Plesk.php:645]: Nevermind.. I realize now they were already there and had just been renamed, but I didn't notice that because the file had been shifted around and the comparison view was a bit off

**Comment by wolveix** [comment]
> I also remain very hesitant about about the addition of the typehints on the properties of this older code. I've been burned far too many times by unexpected behavior from BoxBilling code that breaks when we've added typehints. Have you had the chance to actually that it behaves correctly with those modifications?

Yeah, go figure. No, I haven't tested it properly yet. I was _hoping_ that the existing tests may be sufficient, though perhaps that's some naive trust (given your comments in a previous PR). I generally deploy with Docker, is there a simple way to deploy/test development builds? When I was working on the `passwordLength` implementation, I was just symlinking the `modules` and `library` directories to the container 😅 

> I'm seeing stuff that will for sure be changed with the CS & coding standards workflow. If you wanted to go above-and-beyond you can run Rector and then PHP-CS-Fixer to get those changes in now rather than waiting till the next run of the workflow to create a separate PR for it.

Appreciate knowing those two, I'll run those before all future commits :) 

> Ah thanks. Given the additional context, we should probably perform an `is_numeric` check and then use `intval` or `floatval` (Not sure if there'd be reason for one of these to be a float?) before performing the multiplication.

Good idea, I'll get that sorted.

File: src/library/Server/Manager/Plesk.php
Comment (wolveix) [reply src/library/Server/Manager/Plesk.php:658]: Ah. Yeah, I'm still dialing in my IDE's configuration for PHP, which probably hasn't helped

**Comment by wolveix** [comment]
I wanted to incorporate the string to int conversion within the getters; however, then it'd break the string return. We _could_ make it a `int|string` return, and handle it there? Or add an additional getter (e.g. `getBandwidthBytes`) that solely returns ints?

**Comment by BelleNottelling** [comment]
> I was _hoping_ that the existing tests may be sufficient, though perhaps that's some naive trust (given your comments in a previous PR).

Honestly.. it's not. PHPStan does a really great job of doing a static analysis and will catch *most* issues, but the PHPUnit tests are just terrible. There's nothing that validates the behavior against an actual install with a real database and that's where I've seen issues typically crop up

PRs #1641, #1648, and #1656 are great examples (in order of what happened) of where adding typehints resulting in passing tests, but then a complete failure when it came time to actually write data to the database.

> is there a simple way to deploy/test development builds?

Honestly I don't use docker so I have no clue. @admdly may have some insight though as AFAIK he uses Docker for dev & testing.

> Appreciate knowing those two, I'll run those before all future commits :)

That's probably a tad overkill & should be done with a *hint* of caution. Rector automates refactoring code and on (rare) occasions will accidentally make a change that doesn't work as it should. If you do so, I would suggest making your changes and then running it on a separate commit so it's easy to compare & fix anything if needed. It is quite rare, though. In the ~5k lines of changes it's done to our codebase I think I've only had to manually intervene twice.


> I wanted to incorporate the string to int conversion within the getters; however, then it'd break the string return. We _could_ make it a `int|string` return, and handle it there? Or add an additional getter (e.g. `getBandwidthBytes`) that solely returns ints?

Honestly for all I know the DB might always be returning it as a string so.. *shrugs*. If I'm not mistaken `int` would technically be considered valid in that case, but only as long as we aren't enabling strict types which we probably should be doing for new code and I'd expect it to become more common as things get refactored.

**Comment by wolveix** [comment]
Yeah, I need to figure out some kind of environment to test in. It seems like there are critical setup steps that I may be missing? When trying to add a new hosting server with my current frankenstein setup, I get a `SQLSTATE[42S22]: Column not found: 1054 Unknown column 'password_length' in 'field list' (9999)`. Given that `password_length` doesn't even exist in the codebase, something strange is going on :D

Currently, when replacing the `modules` and `library` directories with my dev versions (and running `Apply Patches & Update Configuration`), the FOSSBilling interface reports that I'm running version `0.0.1`. Ignoring Docker, what does your dev enviornment look like @BelleNottelling? As in, what should the steps be to take the code from my branch, and get up and running with a fresh FOSSBilling instance?

**Comment by BelleNottelling** [comment]
> I get a SQLSTATE[42S22]: Column not found: 1054 Unknown column 'password_length' in 'field list' (9999). Given that password_length doesn't even exist in the codebase, something strange is going on **:D**

Is it possible that your test instance still has some old files from an old version of your other PR?

> FOSSBilling interface reports that I'm running version `0.0.1`

That's typical, the 0.0.1 is the placeholder value in the code.

> As in, what should the steps be to take the code from my branch, and get up and running with a fresh FOSSBilling instance?

So long as you already have a functioning install (so assuming your last PR didn't break anything), it should be as simple as copying any changes over

**Comment by BelleNottelling** [comment]
Ah right.. I just remembered.
If you're wanting a prepped & ready zip of your changes, one will automatically be built so long as the CI is running (so the PR isn't out of date). You can download it from the "Create Preview Release" check artifact. Does take a minute though

**Comment by wolveix** [comment]
@BelleNottelling ah that's great to know, thanks!

Upon testing the latest code in the repo (not from this PR), it seems that the `password_length` error is a legitimate bug. i was a little stumped as `password_length` doesn't exist in the codebase, and I never formatted it as such. After looking into it, it seems that the `$this->di['db']->store($model);` code automatically converts camel case variables like `passwordLength` into `password_length`. I tested this by only replacing `accesshash` with `accessHash` in the respective function, and the same thing happens.

This is called from `src/modules/Servicehosting/Service.phpL684`. I'm going to adjust the database field to use snake case, which should address the issue for now

**Comment by wolveix** [comment]
Now that that's out of the way, everything appears to be working as expected.

It may already be documented, but if not, it'd be good to have the automatic camel case to snake case conversion mentioned somewhere :D

**Comment by wolveix** [comment]
@BelleNottelling yeah, no kidding! Let me know if you needed me to do anything else to get this PR ready for approval :)

**Comment by BelleNottelling** [comment]
> @BelleNottelling yeah, no kidding! Let me know if you needed me to do anything else to get this PR ready for approval :)

Oh, were you able to test and validate that those added type-hints aren't causing unexpected issues?

**Comment by wolveix** [comment]
@BelleNottelling yep, I didn't run into anything when testing out the various server functions that the changes may effect

**Review by BelleNottelling — APPROVED** [review]
LGTM

**Comment by wolveix** [comment]
@BelleNottelling sorry to tag! As I can't add anyone to review, would you mind doing so? Just conscious of getting the `passwordLength` fix merged before the next release

**Comment by wolveix** [comment]
Also, seems the CI tests are failing after merging main (despite no clear indicator as to why)? GH said there was a conflict with the `hestia` code, I went to resolve it, it showed no conflicts, so I merged 🤷🏻‍♂️

**Comment by BelleNottelling** [comment]
> Also, seems the CI tests are failing after merging main (despite no clear indicator as to why)? GH said there was a conflict with the `hestia` code, I went to resolve it, it showed no conflicts, so I merged 🤷🏻‍♂️

Looks like PHP doesn't consider the newly added HestiaCP `generateUsername` function to be compatible with the one in your PR with the typehints:

> PHP Fatal error:  Declaration of Server_Manager_Hestia::generateUsername($domainName) must be compatible with Server_Manager::generateUsername(string $domain): string in C:\Users\Belle\Documents\GitHub\FOSSBilling\src\library\Server\Manager\Hestia.php on line 78

> Just conscious of getting the `passwordLength` fix merged before the next release

Oh I wouldn't publish a release without it, but yeah I can request a review from the others once that error is resolved

**Review by jaapmarcus — APPROVED** [review]

**Pull request merged**
